### PR TITLE
[13804] Topic, instances and keys

### PR DESCRIPTION
**Pull request by JLBuenoLopez-eProsima**
Include a section in the documentation explaining in detail the instances.

### CODE REVIEW

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 1]
```diff
@@ -0,0 +1,62 @@
+.. include:: ../../../03-exports/aliases-api.include
+
+.. _dds_layer_topic_instances:
+
+Topics, keys and instances
+==========================
+
+Even though, by definition, a Topic corresponds to a single data type, it is possible that several topics may refer to
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:8]:
```suggestion
Even though, by definition, a Topic corresponds to a single data type, it is possible that several topic instances may refer to
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+the same data type.
+Therefore, a Topic identifies data of a single type, ranging from one single instance to a whole collection of instances
+of that given type, as shown in the figure below.
+
+.. figure:: /01-figures/instances.png
+    :align: center
+
+The different instances gathered under the same topic are distinguishable by means of some data fields that form the key
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:16]:
```suggestion
The different instances gathered under the same topic are distinguishable by means of one or more data fields that form the key
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+to that data set.
+The key description has to be indicated to the middleware.
+The rule is simple: different data values with the same key value represent successive values for the same instance,
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:19]:
```suggestion
The rule is simple: different data values with the same key value represent successive data samples for the same instance,
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+while different data values with different keys represent different instances.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:20]:
```suggestion
while different data values with different keys represent different topic instances.
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+If no key is provided, the data set associated with the Topic is restricted to a single instance.
+Please refer to :ref:`dds_layer_topic_keyed_data_types` for more information about how to set the key in
+*eProsima Fast DDS*.
+
+The advantage of using instances instead of creating a new :ref:`dds_layer_publisher_dataWriter`,
+:ref:`dds_layer_subscriber_dataReader`, and :ref:`dds_layer_topic_topic` is that the corresponding entity is already
+created and discovered.
+Consequently, there is less memory usage and no new discovery (with the related metatraffic involved as explained in
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:28]:
```suggestion
Consequently, there is less memory usage, and no new discovery (with the related metatraffic involved as explained in
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+:ref:`discovery`) is necessary.
+Another advantage is that some QoS are applied per-instance.
+For example, the :ref:`historyqospolicy` is kept for each instance in the DataWriter.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:31]:
```suggestion
Another advantage is that several QoS are applied per Topic instance; e.g. the :ref:`historyqospolicy` is kept for each instance in the DataWriter.
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+The different instances gathered under the same topic are distinguishable by means of some data fields that form the key
+to that data set.
+The key description has to be indicated to the middleware.
+The rule is simple: different data values with the same key value represent successive values for the same instance,
+while different data values with different keys represent different instances.
+If no key is provided, the data set associated with the Topic is restricted to a single instance.
+Please refer to :ref:`dds_layer_topic_keyed_data_types` for more information about how to set the key in
+*eProsima Fast DDS*.
+
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:24]: I'dd add a title `Instance advantages`

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+The advantage of using instances instead of creating a new :ref:`dds_layer_publisher_dataWriter`,
+:ref:`dds_layer_subscriber_dataReader`, and :ref:`dds_layer_topic_topic` is that the corresponding entity is already
+created and discovered.
+Consequently, there is less memory usage and no new discovery (with the related metatraffic involved as explained in
+:ref:`discovery`) is necessary.
+Another advantage is that some QoS are applied per-instance.
+For example, the :ref:`historyqospolicy` is kept for each instance in the DataWriter.
+Thus, instances could be tuned to a wide range of applications.
+
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:33]: Move examples to separate sections after lifecycle

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+For example, instances could be used to track commercial flights.
+The key identifying the instances could be the airline name and the flight number and the sample the location of each
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:35]:
```suggestion
The key identifying the instances could be the airline name and the flight number, while the sample data would be the location of each
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+flight being tracked.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:36]:
```suggestion
flight being tracked at any given time.
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+Instead of creating a DataWriter, DataReader and Topic each time a flight takes off, the same DDS entities may be reused
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:37]: I'd elaborate that the plane would have the DataWriter and the control tower a single DataReader managing several instances (one per flight)

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+using this DDS standard feature.
+Once the flight has landed, the instance can be disposed (no updates are expected); and if the flight is returned to the
+hangar, the instance is unregistered.
+
+Instances could also be used as a relational database.
+The instance key is analogous to the primary key in a database (unique identifier of something within your data).
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:43]:
```suggestion
The instance key is analogous to the primary key in the database (unique identifier of something within the data).
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+Writing a new sample to the instance is similar to insert/update the database; and disposing the instance is like
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:44]:
```suggestion
In this scenario, writing a new sample to the instance is similar to insert/update the database; and disposing the instance is like
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+removing the data corresponding to the given primary key from the database.
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:45]:
```suggestion
deleting the data corresponding to the given primary key from the database.
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Coordinates
+        double latitude;
+        double longitude;
+        double altitude;
+    };
+
+Then, writing to an specific instance will be done as below:
+
+.. literalinclude:: /../code/DDSCodeTester.cpp
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:92]: Please de-indent

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+The following IDL may define the data model of this specific example:
+
+.. code-block:: idl
+
+    struct FlightPosition
+    {
+        // Unique ID: airline name
+        @key
+        string<256> airline_name;
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:79]:
```suggestion
        @key string<256> airline_name;
```

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+
+        // Unique ID: flight number
+        int flight_number;
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:82]: I'd put this as part of the key as well

**Review by EduPonz — CHANGES_REQUESTED** [review]

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 18]
```diff
@@ -91,6 +90,7 @@ Then, writing to an specific instance will be done as below:
 
 .. literalinclude:: /../code/DDSCodeTester.cpp
    :language: c++
+   :dedent: 8
```
Comment (EduPonz) [inline docs/fastdds/dds_layer/topic/instances.rst:93]: Please look at the generated HTML
```suggestion
   :dedent: 12
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Review by MiguelCompany — APPROVED** [review]

File: docs/fastdds/dds_layer/topic/instances.rst [hunk 19]
```diff
@@ -0,0 +1,100 @@
+.. include:: ../../../03-exports/aliases-api.include
+
+.. _dds_layer_topic_instances:
+
+Topics, keys and instances
+==========================
+
+Even though, by definition, a Topic corresponds to a single data type, it is possible that several topic instances may
```
Comment (richiware) [inline docs/fastdds/dds_layer/topic/instances.rst:8]: Suggestion:

A Topic is linked to a data type and each data sample related to a Topic could be seen as an update in the information described by the data type. But it is possible to add a logical separation and have several instances of a Topic, and a data sample should be an update of the information about a concrete  instance.

**Pull request merged**